lib/src/tokenizer.dart

```diff
@@ -43,76 +43,76 @@ class HtmlTokenizer implements Iterator<Token> {
 
   /// This reference to the parser is used for correct CDATA handling.
   /// The [HtmlParser] will set this at construction time.
-  HtmlParser parser;
+  HtmlParser? parser;
 
-  final Queue<Token> tokenQueue;
+  final Queue<Token?> tokenQueue;
 
   /// Holds the token that is currently being processed.
-  Token currentToken;
+  Token? currentToken;
 
   /// Holds a reference to the method to be invoked for the next parser state.
-  bool Function() state;
+  late bool Function() state;
 
   final StringBuffer _buffer = StringBuffer();
 
-  int _lastOffset;
+  late int _lastOffset;
 
   // TODO(jmesserly): ideally this would be a LinkedHashMap and we wouldn't add
   // an item until it's ready. But the code doesn't have a clear notion of when
   // it's "done" with the attribute.
-  List<TagAttribute> _attributes;
-  Set<String> _attributeNames;
+  List<TagAttribute>? _attributes;
+  Set<String>? _attributeNames;
 
   HtmlTokenizer(doc,
-      {String encoding,
+      {String? encoding,
       bool parseMeta = true,
       this.lowercaseElementName = true,
       this.lowercaseAttrName = true,
       this.generateSpans = false,
-      String sourceUrl,
+      String? sourceUrl,
       this.attributeSpans = false})
       : stream =
             HtmlInputStream(doc, encoding, parseMeta, generateSpans, sourceUrl),
         tokenQueue = Queue() {
     reset();
   }
 
-  TagToken get currentTagToken => currentToken as TagToken;
-  DoctypeToken get currentDoctypeToken => currentToken as DoctypeToken;
-  StringToken get currentStringToken => currentToken as StringToken;
+  TagToken? get currentTagToken => currentToken as TagToken?;
+  DoctypeToken? get currentDoctypeToken => currentToken as DoctypeToken?;
```

Comment:
*All* uses of `currentDoctypeToken` is followed by `!`.
Just do `as DoctypeToken` without the `?` and return  a non-nullable type.

Same for `currentStringToken` and `currentTagToken`.

Reply:
Good call.